### PR TITLE
Add keep_values gap policy

### LINKED ISSUE #27377: Bucket script aggregation returns invalid value for missing docs
**Elasticsearch version** (`bin/elasticsearch --version`): 6.0.0-rc1

**JVM version** (`java -version`):
java version "1.8.0_151"
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)
**OS version** (`uname -a` if on a Unix-like system):
Linux elasticsearch-data-hot-003 4.11.0-1013-azure #13-Ubuntu SMP Mon Oct 2 17:59:06 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:
When doing a `bucket script` aggregation that depends on a cumulative sum aggregation of another sum aggregation, if the sum aggregation returns null values (Because there are no docs in that time interval bucket), the bucket script aggregation will also return null, instead of relying on the cumulative sum value that was gathered so far.

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1. Add docs that span over 5 minutes that look like this:
```js
{
  @timestamp: '',
  bytes: 100
}
```
 2. Run a query that spans **after** the 5m end (meaning that there will be date histogram buckets without docs), with this aggregation:
```json
{
"aggs": {
    "timeseries": {
      "date_histogram": {
        "field": "@timestamp",
        "interval": "1m",
        "min_doc_count": 0,
        "time_zone": "UTC"
      },
      "aggs": {
        "sum_bytes": {
          "sum": {
            "field": "bytes"
          }
        },
        "cumulative_bytes": {
          "cumulative_sum": {
            "buckets_path": "sum_bytes"
          }
        },
        "bucket": {
          "bucket_script": {
            "buckets_path": {
              "bytes": "cumulative_bytes"
            },
            "script": {
              "source": "params.bytes",
              "lang": "painless"
            }
          }
        }
      }
    }
  }
}
```

 3. Check the response and see that in the date histogram without buckets, the `bucket` aggregation does not show the value that its supposed to (The `cumulative_bytes` value)

### PR DESCRIPTION
**Pull request by imotov**
Adds a new keep_values gap policy that works like skip, except if the metric
calculated on an empty bucket provides a non-null non-NaN value, this value is
used for the bucket.

Fixes #27377

### CODE REVIEW

**Comment by imotov** [comment]
@elasticmachine update branch

**Comment by imotov** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by imotov** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by imotov** [comment]
@elasticmachine update branch

**Comment by imotov** [comment]
@elasticmachine update branch

**Comment by imotov** [comment]
@elasticmachine update branch

**Comment by imotov** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketHelpers.java [hunk 1]
```diff
@@ -41,9 +41,40 @@
      *
      * "insert_zeros": empty buckets will be filled with zeros for all metrics
      * "skip": empty buckets will simply be ignored
+     * "none": for empty buckets the values provided by the metrics will still be used if they are available
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketHelpers.java:44]:
```suggestion
     * "keep_values": for empty buckets the values provided by the metrics will still be used if they are available
```
Also, we could make these `{@link }` tags, then renaming would catch them.

**Review by not-napoleon — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketHelpers.java (hunk 1 above)
Comment (imotov) [reply server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketHelpers.java:44]: "none" was never released, it is a leftover from my earlier iteration.

**Comment by imotov** [comment]
@elasticmachine update branch

**Comment by imotov** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Pull request merged**